docs/src/main/tut/TypedDataset.md

```diff
@@ -1 +1,158 @@
+case class Foo(i: Long, j: String)
+
+// Assuming spark is loaded and SparkSession is bind to spark
+val initialDs = spark.createDataset( Foo(1, "Q") :: Foo(10, "W") :: Foo(100, "E") :: Nil )
+
+// Assuming you are on Linux or Mac OS
+initialDs.write.parquet("/tmp/foo")
+
+val ds = spark.read.parquet("/tmp/foo").as[Foo]
+
+ds.show()
+```
+
+The value `ds` holds the content of the `initialDs` read from a parquet file. 
+Let's try to only use field `i` from Foo and see how Spark's Catalyst (the query optimizer) 
+optimizes this. 
+
+```tut:book
+// Using a standard Spark TypedColumn in select()
+val filteredDs = ds.filter($"i" === 10).select($"i".as[Long])
+
+filteredDs.show()
+```
+
+The `filteredDs` is of type `Dataset[Long]`. Since we only access field `i` from `Foo` the type is correct.
+Unfortunately, this syntax requires handholding by explicitly setting the `TypedColumn` in the `select` statement
+to return type `Long` (look at the `as[Long]` statement). We will discuss this limitation next in more detail. 
+Now, let's take a quick look at the optimized Physical Plan that Spark's Catalyst generated.
+
+```tut:book
+
+filteredDs.explain()
+
+```
+
+The last line is very important (see `ReadSchema`). The schema read 
+from the parquet file only required reading column `i` without needing to access column `j`.
+This is great! We have both an optimized query plan and typesafety! 
+
+Unfortunately, this syntax is not bulletproof. 
+This syntax will fail if we try to access a non existing column `x`. 
+The following statement works great at compile time and it will 
+fail at run time (apologies for the long stack trace):
+
+```tut:fail
+ds.filter($"i" === 10).select($"x".as[Long])
+```
+
+There are two things to improve here. Fist, we would want to avoid the `at[Long]` casting that we are required
+to type for typesafely. This is clearly an area where we can introduce a bug but casting to an incompatible 
+type. Second, we want a solution where reference to a 
+non existing column name will fail at compilation time. 
+The standard Spark Dataset can achieve this using the following syntax.
+ 
+```tut:book
+ 
+ds.filter(_.i == 10).map(_.i).show()
+
+```
+
+This looks great! It reminds us the familiar syntax from Scala. 
+The two closures in filter and map are functions that operate on `Foo` and the 
+compiler will helps us capture all the mistakes we mentioned above.
+
+```tut:fail
+ 
+ds.filter(_.i == 10).map(_.x).show()
+
+```
+
+Unfortunately, this syntax does not allow Spark to optimize the code. 
+
+
+```tut:book
+
+ds.filter(_.i == 10).map(_.i).explain()
+
+```
+
+As we see from the explained Physical Plan, Spark was not able to optimize our query as before.
+ Reading the parquet file will required loading all the fields of `Foo`. This might be ok for
+ small datasets or for datasets with few columns, but will be extremely slow for most practical
+ applications. The reason behind this behavior is beyond the scope of this tutorial. 
```

Comment:
I am just going to nuke this "The reason behind this behavior is beyond the scope of this tutorial. "